### DISCUSSION TITLE
Is it possible to use firebase cloud messaging inside pocketbase hooks?

### DISCUSSION THREAD

**Discussion by LimaneGaya** [pocketbase/pocketbase#4402] [Q&A]
I'd like to send notification to users whenever a new row is created but i can't figure out if there is a way to use FCM there.
I tried to use REST API but so far i keep getting errors from being unauthenticated. 
is there a way to use Firebase admin sdk maybe?

**Comment by ganigeorgiev** [comment]
> is there a way to use Firebase admin sdk maybe?

Only if you use Go.

> I tried to use REST API but so far i keep getting errors from being unauthenticated.

Yes, using the Firebase web APIs and `$http.send` is one way to integrate it with the JS `pb_hooks`.
Just make sure that you are sending the proper request data.
To rule out PocketBase issue you could also try sending your request with curl, Postman, Insomnia, etc.
If you need help with this, please provide a code sample showing how you are sending the request.

**Reply by LimaneGaya** [reply]
here is the code i used.
```
const res = $http.send({
            url: "https://fcm.googleapis.com/v1/projects/myprojectid/messages:send",
            method: "POST",
            headers: {
                "Authorization": "Bearer **************************************************",
                "Content-Type": "application/json"
            },
            body: JSON.stringify({
                "message": {
                    "token": user.publicExport()["token"],
                    "notification": {
                        "body": "This is an FCM notification message!",
                        "title": "FCM Message"
                    }
                }
            }), 
        })
        console.log(res.raw)
```
I'm not sure if there is something wrong here, the tokens are fetched correctly and got the autorisation key from https://console.cloud.google.com/iam-admin/serviceaccounts/details/

and keep getting:

```
2024/02/22 13:57:59 {
  "error": {
    "code": 401,
    "message": "Request had invalid authentication credentials. Expected OAuth 2 access token, login cookie or other valid authentication credential. See https://developers.google.com/identity/sign-in/web/devconsole-project.",
    "status": "UNAUTHENTICATED"
  }
}
```
I tried postman too but it's the same thing, maybe i have to use firebase admin sdk.

**Reply by ganigeorgiev** [reply]
Your issue is most likely with the `Authorization` header key - it is not generated properly or doesn't have the necessary permissions, but I don't think I'll be able to help you with this.

> maybe i have to use firebase admin sdk.

The SDK is a wrapper around the API and if you are going to use the same auth key then probably you'll get the same error.

**Reply by ganigeorgiev** [reply]
Maybe this could help https://firebase.google.com/docs/cloud-messaging/auth-server#provide-credentials-manually.

And from the section below it:

> Unless you are using the [Admin SDK](https://firebase.google.com/docs/cloud-messaging/server#firebase-admin-sdk-for-fcm), which handle authorization automatically, you'll need to mint the access token and add it to send requests.

**Comment by mukhtharcm** [comment]
@LimaneGaya did you got any solution?

**Reply by LimaneGaya** [reply]
I managed to make it work back then using legacy FCM API and Pocketbase JS hooks, unfortunately the legacy FCM API stopped working back in July.
I haven't touched my project since long time so i don't have a solution for it. From a quick search i think the only way to do it now is to [build Pocketbase](https://pocketbase.io/docs/go-overview/#getting-started) from source with [Firebase Admin SDK](https://firebase.google.com/docs/admin/setup#add-sdk). You could call the sdk from one of the [Pocketbase event hooks](https://pocketbase.io/docs/go-event-hooks) with one of the [FCM samples](https://firebase.google.com/docs/cloud-messaging/send-message#go).

**Reply by mukhtharcm** [reply]
I've extended the pocketbase with go, and used the Admin SDK.

Like you said, there's no any other way :)
